### PR TITLE
SONAR-7444 SONAR-7396 Identity Provider improvments

### PR DESCRIPTION
**Pull request by julienlancelot**
SONAR-7444 Provide an UnauthorizedException to display functional error
SONAR-7396 Improve error when base URL is badly configured

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/server/authentication/UnauthorizedException.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    super(message, cause);
+  }
+
+  public String getPath() {
+    return format(UNAUTHORIZED_PATH_WITH_MESSAGE, encodeMessage(getMessage()));
+  }
+
+  private static String encodeMessage(String message) {
+    try {
+      return encode(message, Charsets.UTF_8.name());
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/authentication/UnauthorizedException.java]: MINOR Remove this use of "encode"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by teryk** [comment]
Concerning SONAR-7444, I find the message good but it should clearer IMO. `Fail to callback authentication with ''` especially is not clear.
Message in logs:

> 2016.03.15 18:26:16 ERROR web[o.s.s.a.AuthenticationError] Fail to callback authentication with ''
> java.lang.IllegalArgumentException: A valid identity provider key is required. Please check that property 'sonar.core.serverBaseURL' is valid.
> 	at org.sonar.server.authentication.OAuth2CallbackFilter.extractKeyProvider(OAuth2CallbackFilter.java:86) ~[sonar-server-5.5-SNAPSHOT.jar:na]
> 	at org.sonar.server.authentication.OAuth2CallbackFilter.doFilter(OAuth2CallbackFilter.java:64) ~[sonar-server-5.5-SNAPSHOT.jar:na]
> 	at org.sonar.server.platform.MasterServletFilter$GodFilterChain.doFilter(MasterServletFilter.java:125) [sonar-server-5.5-SNAPSHOT.jar:na]
> 	at org.sonar.server.platform.MasterServletFilter.doFilter(MasterServletFilter.java:94) [sonar-server-5.5-SNAPSHOT.jar:na]
> 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.sonar.server.user.UserSessionFilter.doFilter(UserSessionFilter.java:59) [sonar-server-5.5-SNAPSHOT.jar:na]
> 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.sonar.server.platform.ProfilingFilter.doFilter(ProfilingFilter.java:84) [sonar-server-5.5-SNAPSHOT.jar:na]
> 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at ch.qos.logback.access.tomcat.LogbackValve.invoke(LogbackValve.java:191) [logback-access-1.1.3.jar:na]
> 	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:521) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1096) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_66]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_66]
> 	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-8.0.30.jar:8.0.30]
> 	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_66]
>

**Comment by teryk** [comment]
Code LGTM, just need to discuss about error message in log

**Comment by teryk** [comment]
SONAR-7444 LGTM

**Comment by julienlancelot** [comment]
Concerning SONAR-7396, I'll change "Fail to callback authentication with ''" to "Fail to callback authentication".
But I don't want to display to value of the property sonar.core.serverBaseURL, as it will force me to inject Settings in OAuth2CallbackFilter only for that, and moreover as the user needs to change this value he must go to the settings page to update it.

**Comment by teryk** [comment]
that's fine: LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**